### DISCUSSION TITLE
Looking for help with versioning in a filesystem registry

### DISCUSSION THREAD

**Discussion by Bvsemmer** [microsoft/vcpkg#38710] [Q&A] [answered]
Hey, 
I am doing experiments with vcpkg, to potentially use it as a C++ package manager. I am setting up a test environment to try some things out. Though, I fail to get a certain workflow, I have in mind, working. I can use some help.

The absolute requirement for me is that is uses a filesystem registry. Not Git based.

After reading the documentation, I have a working environment where I have a local port of a certain library. Installing the package works fine and the app project builds and links correctly.

The problem is that I cannot get the versioning to work with a filesystem registry and the documentation is not clear about it. At least, I am not getting results I expect. 

So, in short. I have the following package 'depot' structure on my local filesystem

```
/
├── dummylib
│   └── 1.2.3
│       └── <src>
├── ports
│   ├── dummylib
│   │   ├── portfile.cmake
│   │   └── vcpkg.json
│   └── tinyxml2
│       ├── portfile.cmake
│       └── vcpkg.json
├── tinyxml2
│   └── 10.0.0
│       └── <src>
└── versions
    ├── d-
    │   └── dummylib.json
    └── t-
        └── tinyxml2.json
```

My app folder structure looks like this and is located outside of this 'depot' mentioned above:
```
/
├── CMakeLists.txt
├── CMakePresets.json
├── vcpkg.json
├── vcpkg-configuration.json
└── <src>
```

In vcpkg-configuration.json, I added the 'overlay' path
`
"overlay-ports": [
    "<depot root>/ports"
  ]
`

In vcpkg.json, I added the following dependency
`
 "dependencies": [
    {
      "name": "tinyxml2",
      "version>=": "11.0.0"
    }
  ]
`

I run `vcpkg install` in the app folder to install all dependencies. What I would expect here is that vcpkg would FAIL in this case. Since I ask for tinyxml2 version '11.0.0', while I only have '10.0.0' in my registry. This is not happening, it just installs whatever tinyxml2 version it finds locally in the ports/ folder. 
This behaviour is different compared to installing tinyxml without an overlay port. In that case, I DO get an error when I make a dependency on a non existing version.

Looking at how versioning is explained here (https://learn.microsoft.com/en-us/vcpkg/users/versioning), I fail to see how vcpkg even knowns about the `versions/` folder in my 'depot'. I see no way of making vcpkg aware of that. And I cannot find any documentation about it. 

Thanks a lot in advance!

**Comment by Neumann-A** [comment, marked as answer]
> Not Git based.

Why? git stuff does not need to be remote. You can have totally local git trees. 

> What I would expect here is that vcpkg would FAIL in this case. Since I ask for tinyxml2 version '11.0.0', while I only have '10.0.0' in my registry.

overlays are not registries. overlays are the ultimate override. They will override any version specifications. You want to use a filesystem registry not an overlay. (see https://github.com/vcpkg/example-filesystem-registry)

As such you need something like the following in your `vcpkg-configuration.json`
```
  "registries": [
    {
      "kind": "filesystem",
      "repository": "<path-to_registry>"
   }
```

**Comment by Bvsemmer** [comment]
Thank you for your answer! 
I've seen that filesystem register example indeed, but it does not include an example on how to configure your actual app with such a registry unfortunately. 

I've tried modifying `vcpkg-configuration.json`, as follows (using 'repository' instead of 'path', which gave me an error)
```
"registries": [
    {
      "kind": "filesystem",
      "path": "<path to depot containing ports/ and versions/>",
      "packages": [ "tinyxml2" ]
    }
  ]
```

That fixed my issue, thanks a lot!

> Why? git stuff does not need to be remote. You can have totally local git trees.

Simply because our package source is not stored in a Git repository.

**Reply by dg0yt** [reply]
> > Why? git stuff does not need to be remote. You can have totally local git trees.
> 
> Simply because our package source is not stored in a Git repository.

Your package source doesn't need to live in git in order to use vcpkg git registries.
